File: VBNetCore.Mvvm/Extensions/ServiceCollectionExtension.vb
Imports Microsoft.Extensions.DependencyInjection
Imports VBNetCore.Mvvm.Abstractions
Imports VBNetCore.Mvvm.Controls
Imports VBNetCore.Mvvm.Helpers
Imports VBNetCore.Mvvm.ViewModels
Imports System.Linq
Imports System.Windows
Imports System.Runtime.CompilerServices

Namespace VBNetCore.Mvvm.Extensions
    ''' <summary>
    ''' IServiceCollection extension methods for common scenarios.
    ''' </summary>
    Public Module ServiceCollectionExtension
        ''' <summary>
        ''' Adds required services to the given service collection.
        ''' </summary>
        ''' <paramname="services">The service collection.</param>
        ''' <returns>A reference to this instance after the operation has completed.</returns>
        <Extension()>
        Public Function AddMvvm(ByVal services As IServiceCollection) As IServiceCollection
            services.AddSingleton(Of ViewLocator)()
            services.AddSingleton(Of IWindowManager, WindowManager)()
            services.AddSingleton(Of IPopupManager, PopupManager)()
            services.AddSingleton(Of IUiExecution, UiExecution)()
            services.AddMvvmSingleton(Of PopupMessageViewModel, PopupMessageView)()
            Return services
        End Function

        ''' <summary>
        ''' Adds a transient viewmodel and its corresponding view to the service collection.
        ''' </summary>
        ''' <paramname="services">The <seecref="IServiceCollection"/> to add the types to.</param>
        ''' <typeparamname="TViewModel">The type of the viewmodel to add.</typeparam>
        ''' <typeparamname="TView">The type of the view to add.</typeparam>
        ''' <returns>A reference to this instance after the operation has completed.</returns>
        <Extension()>
        Public Function AddMvvmTransient(Of TViewModel As Class, TView As FrameworkElement)(ByVal services As IServiceCollection) As IServiceCollection
            services.AddTransient(Of TViewModel)().AddTransient(Of TView)()
            services.GetViewModelRegistry().ViewModelTypeToViewType.Add(GetType(TViewModel), GetType(TView))

            Return services
        End Function

        ''' <summary>
        ''' Adds a singleton viewmodel and its corresponding view to the service collection.
        ''' </summary>
        ''' <paramname="services">The <seecref="IServiceCollection"/> to add the types to.</param>
        ''' <typeparamname="TViewModel">The type of the viewmodel to add.</typeparam>
        ''' <typeparamname="TView">The type of the view to add.</typeparam>
        ''' <returns>A reference to this instance after the operation has completed.</returns>
        <Extension()>
        Public Function AddMvvmSingleton(Of TViewModel As Class, TView As FrameworkElement)(ByVal services As IServiceCollection) As IServiceCollection
            services.AddSingleton(Of TViewModel)().AddTransient(Of TView)()
            services.GetViewModelRegistry().ViewModelTypeToViewType.Add(GetType(TViewModel), GetType(TView))

            Return services
        End Function

        <Extension()>
        Private Function GetViewModelRegistry(ByVal services As IServiceCollection) As ViewModelRegistry
            Dim registry As ViewModelRegistry
            Dim descriptor = services.FirstOrDefault(Function(x) x.ServiceType = GetType(ViewModelRegistry))
            If descriptor IsNot Nothing Then
                registry = CType(descriptor.ImplementationInstance, ViewModelRegistry)
            Else
                registry = New ViewModelRegistry()
                services.AddSingleton(registry)
            End If

            Return registry
        End Function
    End Module
End Namespace

File: VBNetCore.Mvvm/Extensions/ServiceCollectionExtension.cs
using Microsoft.Extensions.DependencyInjection;
using NetCore.Mvvm.Abstractions;
using NetCore.Mvvm.Controls;
using NetCore.Mvvm.Helpers;
using NetCore.Mvvm.ViewModels;
using System.Linq;
using System.Windows;

namespace NetCore.Mvvm.Extensions
{
    /// <summary>
    /// IServiceCollection extension methods for common scenarios.
    /// </summary>
    public static class ServiceCollectionExtension
    {
        /// <summary>
        /// Adds required services to the given service collection.
        /// </summary>
        /// <param name="services">The service collection.</param>
        /// <returns>A reference to this instance after the operation has completed.</returns>
        public static IServiceCollection AddMvvm(this IServiceCollection services)
        {
            services.AddSingleton<ViewLocator>();
            services.AddSingleton<IWindowManager, WindowManager>();
            services.AddSingleton<IPopupManager, PopupManager>();
            services.AddSingleton<IUiExecution, UiExecution>();
            services.AddMvvmSingleton<PopupMessageViewModel, PopupMessageView>();
            return services;
        }

        /// <summary>
        /// Adds a transient viewmodel and its corresponding view to the service collection.
        /// </summary>
        /// <param name="services">The <see cref="IServiceCollection"/> to add the types to.</param>
        /// <typeparam name="TViewModel">The type of the viewmodel to add.</typeparam>
        /// <typeparam name="TView">The type of the view to add.</typeparam>
        /// <returns>A reference to this instance after the operation has completed.</returns>
        public static IServiceCollection AddMvvmTransient<TViewModel, TView>(this IServiceCollection services) where TViewModel : class where TView : FrameworkElement
        {
            services
                .AddTransient<TViewModel>()
                .AddTransient<TView>();

            services
                .GetViewModelRegistry()
                .ViewModelTypeToViewType.Add(typeof(TViewModel), typeof(TView));

            return services;
        }

        /// <summary>
        /// Adds a singleton viewmodel and its corresponding view to the service collection.
        /// </summary>
        /// <param name="services">The <see cref="IServiceCollection"/> to add the types to.</param>
        /// <typeparam name="TViewModel">The type of the viewmodel to add.</typeparam>
        /// <typeparam name="TView">The type of the view to add.</typeparam>
        /// <returns>A reference to this instance after the operation has completed.</returns>
        public static IServiceCollection AddMvvmSingleton<TViewModel, TView>(this IServiceCollection services) where TViewModel : class where TView : FrameworkElement
        {
            services
                .AddSingleton<TViewModel>()
                .AddTransient<TView>();

            services
                .GetViewModelRegistry()
                .ViewModelTypeToViewType.Add(typeof(TViewModel), typeof(TView));

            return services;
        }

        private static ViewModelRegistry GetViewModelRegistry(this IServiceCollection services)
        {
            ViewModelRegistry registry;
            var descriptor = services.FirstOrDefault(x => x.ServiceType == typeof(ViewModelRegistry));
            if (descriptor != null)
            {
                registry = (ViewModelRegistry)descriptor.ImplementationInstance!;
            }
            else
            {
                registry = new ViewModelRegistry();
                services.AddSingleton(registry);
            }

            return registry;
        }
    }
}
